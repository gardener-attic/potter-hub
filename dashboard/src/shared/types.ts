import * as jsonSchema from "json-schema";
import { IAuthState } from "../reducers/auth";
import { IClusterBomState } from "../reducers/clusterbom";
import { IConfigState } from "../reducers/config";
import { INamespaceState } from "../reducers/namespace";
import { IAppRepositoryState } from "../reducers/repos";
import { IServiceCatalogState } from "../reducers/serviceCatalog";
import { hapi } from "./hapi/release";
import { ResourceKind } from "./ResourceKinds";

// Allow defining multiple error classes
// tslint:disable:max-classes-per-file

class CustomError extends Error {
  // The constructor is defined so we can later on compare the returned object
  // via err.contructor  == FOO
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
export class ForbiddenError extends CustomError {}
export class UnauthorizedError extends CustomError {}

export class NotFoundError extends CustomError {}

export class ConflictError extends CustomError {}

export class UnprocessableEntity extends CustomError {}

export interface IRepo {
  name: string;
  url: string;
}

export interface IChartVersion {
  id: string;
  attributes: IChartVersionAttributes;
  relationships: {
    chart: {
      data: IChartAttributes;
    };
  };
}

export interface IChartVersionAttributes {
  version: string;
  app_version: string;
  created: string;
}

export interface IChart {
  id: string;
  attributes: IChartAttributes;
  relationships: {
    latestChartVersion: {
      data: IChartVersionAttributes;
    };
  };
}

export interface IChartAttributes {
  name: string;
  description: string;
  home?: string;
  icon?: string;
  keywords: string[];
  maintainers: Array<{
    name: string;
    email?: string;
  }>;
  repo: IRepo;
  sources: string[];
}

export interface IChartState {
  isFetching: boolean;
  selected: {
    error?: Error;
    version?: IChartVersion;
    versions: IChartVersion[];
    readme?: string;
    readmeError?: string;
    values?: string;
    schema?: any;
  };
  deployed: {
    chartVersion?: IChartVersion;
    values?: string;
    schema?: jsonSchema.JSONSchema4;
  };
  items: IChart[];
}

export interface IChartUpdateInfo {
  upToDate: boolean;
  chartLatestVersion: string;
  appLatestVersion: string;
  repository: IRepo;
  error?: Error;
}

export interface IDeployment {
  metadata: {
    name: string;
    namespace: string;
  };
}

export interface IServiceSpec {
  ports: IPort[];
  clusterIP: string;
  type: string;
}

export interface IServiceStatus {
  loadBalancer: {
    ingress?: Array<{ ip?: string; hostname?: string }>;
  };
}

export interface IPort {
  name: string;
  port: number;
  protocol: string;
  targetPort: string;
  nodePort: string;
}

export interface IHTTPIngressPath {
  path: string;
}
export interface IIngressHTTP {
  paths: IHTTPIngressPath[];
}
export interface IIngressRule {
  host: string;
  http: IIngressHTTP;
}

export interface IIngressTLS {
  hosts: string[];
}

export interface IIngressSpec {
  rules: IIngressRule[];
  tls?: IIngressTLS[];
}

export interface IResource {
  apiVersion: string;
  kind: ResourceKind;
  type: string;
  spec: any;
  status: any;
  metadata: {
    name: string;
    namespace: string;
    annotations?: { [key: string]: string };
    creationTimestamp: string;
    deletionTimestamp?: string | null;
    generation?: number;
    labels?: { [key: string]: string };
    resourceVersion: string;
    uid: string;
    selfLink: string;
  };
}

export interface IOwnerReference {
  apiVersion: string;
  blockOwnerDeletion: boolean;
  kind: string;
  name: string;
  uid: string;
}

export interface ISecret {
  apiVersion: string;
  kind: string;
  type: string;
  data: { [s: string]: string };
  metadata: {
    name: string;
    namespace: string;
    annotations: string;
    creationTimestamp: string;
    selfLink: string;
    resourceVersion: string;
    deletionTimestamp?: string;
    uid: string;
  };
}

export interface IDeploymentStatus {
  replicas: number;
  updatedReplicas: number;
  availableReplicas: number;
}

export interface IStatefulsetStatus {
  replicas: number;
  updatedReplicas: number;
  readyReplicas: number;
}

export interface IDaemonsetStatus {
  currentNumberScheduled: number;
  numberReady: number;
}

export interface IRelease extends hapi.release.Release {
  updateInfo?: IChartUpdateInfo;
}

export interface IAppState {
  isFetching: boolean;
  error?: Error;
  deleteError?: Error;
  // currently items are always Helm releases
  items: IRelease[];
  listingAll: boolean;
  listOverview?: IAppOverview[];
  selected?: IRelease;
}

export interface IStoreState {
  serviceCatalog: IServiceCatalogState;
  apps: IAppState;
  auth: IAuthState;
  charts: IChartState;
  clusterBom: IClusterBomState;
  config: IConfigState;
  kube: IKubeState;
  repos: IAppRepositoryState;
  namespace: INamespaceState;
}

interface IK8sResource {
  apiVersion: string;
  kind: string;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#objects */
export interface IK8sObject<M, SP, ST> extends IK8sResource {
  metadata: {
    annotations?: { [key: string]: string };
    creationTimestamp?: string;
    deletionTimestamp?: string | null;
    generation?: number;
    labels?: { [key: string]: string };
    name: string;
    namespace: string;
    resourceVersion?: string;
    uid: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
  spec?: SP;
  status?: ST;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#lists-and-simple-kinds */
export interface IK8sList<I, M> extends IK8sResource {
  items: I[];
  metadata?: {
    resourceVersion?: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
}

export interface IAppRepository
  extends IK8sObject<
    {
      clusterName: string;
      creationTimestamp: string;
      deletionGracePeriodSeconds: string | null;
      deletionTimestamp: string | null;
      resourceVersion: string;
      selfLink: string;
    },
    {
      type: string;
      url: string;
      auth: {
        header: {
          secretKeyRef: {
            name: string;
            key: string;
          };
        };
      };
      resyncRequests: number;
    },
    undefined
  > {}

export interface IAppRepositoryList
  extends IK8sList<
    IAppRepository,
    {
      continue: string;
      resourceVersion: string;
      selfLink: string;
    }
  > {}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#response-status-kind */
export interface IStatus extends IK8sResource {
  kind: "Status";
  status: "Success" | "Failure";
  message: string;
  reason:
    | "BadRequest"
    | "Unauthorized"
    | "Forbidden"
    | "NotFound"
    | "AlreadyExists"
    | "Conflict"
    | "Invalid"
    | "Timeout"
    | "ServerTimeout"
    | "MethodNotAllowed"
    | "InternalError";
  details?: {
    kind?: string;
    name?: string;
    causes?: IStatusCause[] | string;
  };
}

interface IStatusCause {
  field: string;
  message: string;
  reason: string;
}

export interface IRouterPathname {
  router: {
    location: {
      pathname: string;
    };
  };
}

export interface IRuntimeVersion {
  name: string;
  version: string;
  runtimeImage: string;
  initImage: string;
}

export interface IRuntime {
  ID: string;
  versions: IRuntimeVersion[];
  depName: string;
  fileNameSuffix: string;
}

export interface IRBACRole {
  apiGroup: string;
  namespace?: string;
  clusterWide?: boolean;
  resource: string;
  verbs: string[];
}

export interface IAppOverview {
  releaseName: string;
  namespace: string;
  version: string;
  icon?: string;
  status: string;
  chart: string;
  chartMetadata: hapi.chart.Metadata;
  // UpdateInfo is internally populated
  updateInfo?: IChartUpdateInfo;
  description: string;
}

export interface IKubeItem<T> {
  isFetching: boolean;
  item?: T;
  error?: Error;
}

export interface IKubeState {
  items: { [s: string]: IKubeItem<IResource> };
  fetchTargetClusterSecretErr?: Error;
  sockets: { [s: string]: { socket: WebSocket; closeTimer: () => void } };
}

export interface IBasicFormParam {
  path: string;
  type?: jsonSchema.JSONSchema4TypeName | jsonSchema.JSONSchema4TypeName[];
  value?: any;
  title?: string;
  minimum?: number;
  maximum?: number;
  render?: string;
  description?: string;
  enum?: string[];
  hidden?:
    | {
        path: any;
        value: string;
        conditions: Array<{
          path: any;
          value: string;
        }>;
        operator: string;
      }
    | string;
  children?: IBasicFormParam[];
}
export interface IBasicFormSliderParam extends IBasicFormParam {
  sliderMin?: number;
  sliderMax?: number;
  sliderUnit?: string;
}

export interface IApplicationConfig {
  id: string;
  configType: string;
  typeSpecificData: {};
  values: {};
  secretValues: {};
  noReconcile: boolean;
  readyRequirements: IReadyRequirements;
}

export interface IReadyRequirements {
    jobs: IJobReadyRequirement[];
    resources: IResourceReadyRequirement[];
}

export interface IJobReadyRequirement {
    name: string;
    namespace: string
}

export interface IResourceReadyRequirement {
    name: string;
    namespace: string;
    apiVersion: string;
    resource: string;
    fieldPath: string;
    successValues: ISuccessValue[];
}

export interface ISuccessValue {
    value: {};
}

export interface IAutoDelete {
    clusterBomAge: number;
}

export interface IClusterBom
  extends IK8sObject<
    {},
    {
      secretRef: string;
      applicationConfigs: IApplicationConfig[];
      autoDelete: IAutoDelete;
    },
    {
      observedGeneration: number;
      overallState: string;
      overallTime: string;
      conditions: ICondition[];
      applicationStates: IApplicationState[];
      overallNumOfDeployments: number;
      overallNumOfReadyDeployments: number;
      overallProgress: number;
    }
  > {}

export interface ICondition {
  type: string;
  status: string;
  lastUpdateTime: string;
  lastTransitionTime: string;
  reason: string;
  message: string;
}

export interface IApplicationState {
  id: string;
  state: string;
  detailedState: {};
}

export interface IClusterBomList extends IK8sList<IClusterBom, {}> {}

export interface IURLParams {
    targetClusterSecretName: string;
    targetClusterSecretNamespace: string;
}
