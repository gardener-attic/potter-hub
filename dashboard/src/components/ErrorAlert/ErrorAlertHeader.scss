.error__icon {
  display: inline-block;
  background: #c7c9c8;
  vertical-align: bottom;
  border-radius: 50%;
  min-height: 30px;
  min-width: 30px;
}

.error__icon > svg {
  margin: auto;
  left: 3px;
  top: 3px;
  color: #ffffff;
  position: relative;
}
