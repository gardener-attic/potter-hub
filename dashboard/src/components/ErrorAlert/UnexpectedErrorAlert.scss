.terminal__error {
  word-break: break-all;
}

.error__text {
  white-space: pre-wrap;
}
