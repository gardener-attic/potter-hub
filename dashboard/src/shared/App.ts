import { axiosWithAuth } from "./AxiosInstance";
import { hapi } from "./hapi/release";
import { HELM_URL } from "./Kube";
import { IAppOverview, IChartVersion } from "./types";

export class App {
  public static getResourceURL(namespace?: string, name?: string, query?: string) {
    let url = HELM_URL;
    if (namespace) {
      url += `/namespaces/${namespace}`;
    }
    url += "/releases";
    if (name) {
      url += `/${name}`;
    }
    if (query) {
      url += `?${query}`;
    }
    return url;
  }

  public static async create(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = App.getResourceURL(namespace);
    const { data } = await axiosWithAuth.post(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async upgrade(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axiosWithAuth.put(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async rollback(releaseName: string, namespace: string, revision: number) {
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axiosWithAuth.put(
      endpoint,
      {},
      {
        params: {
          action: "rollback",
          revision,
        },
      },
    );
    return data;
  }

  public static async delete(releaseName: string, namespace: string, keepHistory: boolean) {
    let keepHistoryQuery;
    if (keepHistory) {
      keepHistoryQuery = "keepHistory=true";
    }
    const { data } = await axiosWithAuth.delete(
      App.getResourceURL(namespace, releaseName, keepHistoryQuery),
    );
    return data;
  }

  public static async listApps(namespace?: string, allStatuses?: boolean) {
    let query;
    if (allStatuses) {
      query = "statuses=all";
    }
    const { data } = await axiosWithAuth.get<{ data: IAppOverview[] }>(
      App.getResourceURL(namespace, undefined, query),
    );
    return data.data;
  }

  public static async getRelease(namespace: string, name: string) {
    const { data } = await axiosWithAuth.get<{ data: hapi.release.Release }>(
      this.getResourceURL(namespace, name),
    );
    return data.data;
  }
}
