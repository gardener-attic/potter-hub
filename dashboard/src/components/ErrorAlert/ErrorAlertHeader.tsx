import * as React from "react";
import { AlertTriangle } from "react-feather";

import "./ErrorAlertHeader.css";

interface IErrorHeaderProps {
  children: string | JSX.Element | Array<string | JSX.Element>;
  icon?: React.Component;
}

class ErrorPageHeader extends React.Component<IErrorHeaderProps> {
  public render() {
    const { children } = this.props;
    const Icon = this.props.icon || AlertTriangle;
    return (
      <header>
        <div className="margin-b-big">
          <h5 className="type-regular">
            <span className="error__icon margin-r-small">
              <Icon />
            </span>
            {children}
          </h5>
        </div>
      </header>
    );
  }
}

export default ErrorPageHeader;
