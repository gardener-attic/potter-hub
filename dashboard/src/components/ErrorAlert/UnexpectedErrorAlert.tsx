import * as React from "react";
import { X } from "react-feather";

import ErrorPageHeader from "./ErrorAlertHeader";
import "./UnexpectedErrorAlert.css";

interface IUnexpectedErrorPage {
  icon?: any;
  raw?: boolean;
  showGenericMessage?: boolean;
  text?: string;
  title?: string | JSX.Element;
}

export const genericMessage = (
  <div>
    <p>Troubleshooting:</p>
    <ul className="error__troubleshooting">
      <li>Check for network issues.</li>
      <li>Check your browser's JavaScript console for additional errors.</li>
      <li>
        <a href="https://github.com/gardener/potter-hub/issues/new/choose" target="_blank" rel="noopener noreferrer">
          Open an issue on GitHub
        </a>{" "}
        if you think you've encountered a bug.
      </li>
    </ul>
  </div>
);

class UnexpectedErrorPage extends React.Component<IUnexpectedErrorPage> {
  public static defaultProps: Partial<IUnexpectedErrorPage> = {
    title: "Sorry! Something went wrong.",
    showGenericMessage: false,
  };
  public render() {
    let customMessage: JSX.Element | null = null;
    if (this.props.text) {
      if (this.props.raw) {
        customMessage = (
          <div className="error__content">
            <section className="Terminal terminal__error elevation-1 type-color-white error__text">
              {this.props.text}
            </section>
          </div>
        );
      } else {
        customMessage = <p>{this.props.text}</p>;
      }
    }
    // NOTE(miguel) We are using the non-undefined "!" token in `props.title` because our current version of
    // typescript does not support react's defaultProps and we are running it in strictNullChecks mode.
    // Newer versions of it seems to support it https://github.com/Microsoft/TypeScript/wiki/Roadmap#30-july-2018
    return (
      <div className="alert alert-error margin-v-bigger">
        <ErrorPageHeader icon={this.props.icon || X}>{this.props.title!}</ErrorPageHeader>
        {customMessage && (
          <div className="error__content margin-l-enormous margin-b-big">{customMessage}</div>
        )}
        {this.props.showGenericMessage && (
          <div className="error__content margin-l-enormous">{genericMessage}</div>
        )}
        {this.props.children && (
          <div className="error__content margin-l-enormous">{this.props.children}</div>
        )}
      </div>
    );
  }
}

export default UnexpectedErrorPage;
