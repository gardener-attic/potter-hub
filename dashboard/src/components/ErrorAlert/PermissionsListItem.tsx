import * as React from "react";
import { IRBACRole } from "shared/types";
import { definedNamespaces } from "../../shared/Namespace";

import { namespaceText } from "./helpers";

interface IPermissionsListItemProps {
  role: IRBACRole;
  namespace: string;
}

class PermissionsListItem extends React.Component<IPermissionsListItemProps> {
  public render() {
    const { role } = this.props;
    const namespace = role.clusterWide
      ? definedNamespaces.all
      : role.namespace || this.props.namespace;
    return (
      <li>
        {role.verbs.join(", ")} <code>{role.resource}</code>{" "}
        {role.apiGroup !== "" ? (
          <span>
            (<code>{role.apiGroup}</code>)
          </span>
        ) : (
          ""
        )}{" "}
        in {namespaceText(namespace)}.
      </li>
    );
  }
}

export default PermissionsListItem;
