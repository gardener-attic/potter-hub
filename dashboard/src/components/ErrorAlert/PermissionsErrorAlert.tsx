import * as React from "react";
import { Lock } from "react-feather";

import { UnexpectedErrorAlert } from ".";
import { IRBACRole } from "../../shared/types";
import { namespaceText } from "./helpers";
import PermissionsListItem from "./PermissionsListItem";

interface IPermissionsErrorPage {
  action: string;
  roles: IRBACRole[];
  namespace: string;
}

class PermissionsErrorPage extends React.Component<IPermissionsErrorPage> {
  public render() {
    const { action, namespace, roles } = this.props;
    return (
      <UnexpectedErrorAlert
        title={
          <span>
            You don't have sufficient permissions to {action} in {namespaceText(namespace)}
          </span>
        }
        icon={Lock}
        showGenericMessage={false}
      >
        <div>
          <p>Ask your administrator for the following RBAC roles:</p>
          <ul className="error__permisions-list">
            {roles.map((r, i) => (
              <PermissionsListItem key={i} namespace={namespace} role={r} />
            ))}
          </ul>
          <p>
            Ask you Gardener project manager to add you as a member, so that you have access to the
            cluster.
          </p>
        </div>
      </UnexpectedErrorAlert>
    );
  }
}

export default PermissionsErrorPage;
