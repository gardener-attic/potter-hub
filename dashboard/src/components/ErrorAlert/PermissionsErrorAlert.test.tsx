import { shallow } from "enzyme";
import * as React from "react";

import { UnexpectedErrorAlert } from ".";
import { IRBACRole } from "../../shared/types";
import ErrorAlertHeader from "./ErrorAlertHeader";
import PermissionsErrorAlert from "./PermissionsErrorAlert";
import PermissionsListItem from "./PermissionsListItem";
import { genericMessage } from "./UnexpectedErrorAlert";

it("renders an error message for the action", () => {
  const roles: IRBACRole[] = [];
  const action = "unit-test";
  const wrapper = shallow(<PermissionsErrorAlert roles={roles} action={action} namespace="test" />);
  const header = wrapper
    .find(UnexpectedErrorAlert)
    .shallow()
    .find(ErrorAlertHeader);
  expect(header).toExist();
  expect(header.shallow().text()).toContain(`You don't have sufficient permissions to ${action}`);
  expect(wrapper.html()).toContain("Ask your administrator for the following RBAC roles:");
  expect(wrapper).toMatchSnapshot();
});

it("renders PermissionsListItem for each RBAC role", () => {
  const roles: IRBACRole[] = [
    {
      apiGroup: "test.kubeapps.com",
      resource: "tests",
      verbs: ["get", "create"],
    },
    {
      apiGroup: "apps",
      namespace: "test",
      resource: "deployments",
      verbs: ["list", "watch"],
    },
  ];
  const wrapper = shallow(<PermissionsErrorAlert roles={roles} action="test" namespace="test" />);
  expect(wrapper.find(PermissionsListItem)).toHaveLength(2);
});

it("renders a information about access control", () => {
  const roles: IRBACRole[] = [];
  const wrapper = shallow(<PermissionsErrorAlert roles={roles} action="test" namespace="test" />);
  expect(wrapper.html()).toMatch(
    /Ask you Gardener project manager to add you as a member, so that you have access to the cluster./,
  );
  expect(wrapper.html()).not.toContain(shallow(genericMessage).html());
});
