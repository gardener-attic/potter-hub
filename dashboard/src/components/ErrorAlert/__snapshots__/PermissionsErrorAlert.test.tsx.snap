// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an error message for the action 1`] = `
<UnexpectedErrorPage
  icon={[Function]}
  showGenericMessage={false}
  title={
    <span>
      You don't have sufficient permissions to 
      unit-test
       in 
      <span>
        the 
        <code>
          test
        </code>
         namespace
      </span>
    </span>
  }
>
  <div>
    <p>
      Ask your administrator for the following RBAC roles:
    </p>
    <ul
      className="error__permisions-list"
    />
    <p>
      Ask you Gardener project manager to add you as a member, so that you have access to the cluster.
    </p>
  </div>
</UnexpectedErrorPage>
`;
